File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.py
import sys
input = sys.stdin.readline

n = int(input().rstrip())
arr = list(map(int, input().rstrip().split()))
dp = [0 for _ in range(1000001)]
size = 1

def binarySearch(now, start, end):
    mid = (start + end) // 2
    if start > end: return -1
    if dp[mid] < now <= dp[mid+1]: return mid+1
    if dp[mid] >= now: return binarySearch(now, start, mid - 1)
    if dp[mid] < now: return binarySearch(now, mid + 1, end)

for a in arr:
    index = binarySearch(a, 0, size-1)

    if index == -1:
        dp[size] = a
        size += 1
    else:
        dp[index] = a

print(size - 1)

File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

public class Main {
	public static void main(String[] args) throws IOException {
		// Input Start
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		int n = Integer.parseInt(br.readLine());
		
		StringTokenizer st = new StringTokenizer(br.readLine());
		int[] arr = new int[n];
		
		for (int i = 0; i < n; i++) {
			arr[i] = Integer.parseInt(st.nextToken());
		}
		
		// Input End
		br.close();
		
		// Solve : LIS with Binary Search
		int[] dp = new int[n];
		int right = 0;
		
		for (int i = 0; i < n; i++) {
			int idx = lower_bound(dp, right, arr[i]);
			right = Math.max(right, idx + 1);
			
			dp[idx] = arr[i];
		}
		
		// Output Start
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		bw.write(right + "\n");
		bw.flush();
		
		// Output End
		bw.close();
		
	}
	
	public static int lower_bound(int[] dp, int r, int k) {
		int l = 0;
		while (l < r) {
			int mid = (l + r) / 2;
			
			if (dp[mid] >= k) r = mid;
			else l = mid + 1;
		}
		
		return l;
	}
}
